File: rocketmq-go/model/message_ext.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package model

import (
	"github.com/apache/incubator-rocketmq-externals/rocketmq-go/model/constant"
	"github.com/apache/incubator-rocketmq-externals/rocketmq-go/util"
	"math"
	"strconv"
	"time"
)

type MessageExt struct {
	*Message
	QueueId                   int32
	StoreSize                 int32
	QueueOffset               int64
	SysFlag                   int32
	BornTimestamp             int64
	BornHost                  string
	StoreTimestamp            int64
	StoreHost                 string
	MsgId                     string
	CommitLogOffset           int64
	BodyCRC                   int32
	ReconsumeTimes            int32
	PreparedTransactionOffset int64

	propertyConsumeStartTimestamp string // race condition
}

func (self *MessageExt) GetOriginMessageId() string {
	if self.Properties != nil {
		originMessageId := self.Properties[constant.PROPERTY_ORIGIN_MESSAGE_ID]
		if len(originMessageId) > 0 {
			return originMessageId
		}
	}
	return self.MsgId
}

func (self *MessageExt) GetConsumeStartTime() int64 {
	if len(self.propertyConsumeStartTimestamp) > 0 {
		return util.StrToInt64WithDefaultValue(self.propertyConsumeStartTimestamp, -1)
	}
	return math.MaxInt64
}

func (self *MessageExt) SetConsumeStartTime() {
	if self.Properties == nil {
		self.Properties = make(map[string]string)
	}
	nowTime := strconv.FormatInt(time.Now().UnixNano()/1000000, 10)
	self.Properties[constant.PROPERTY_KEYS] = nowTime
	self.propertyConsumeStartTimestamp = nowTime
	return
}

File: rocketmq-client-php/src/message_ext.cc
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

#include "message_ext.h"

void registerMessageExt(Php::Namespace &rocketMQNamespace){
    Php::Class<MessageExt> messageExtClass("MessageExt");

    messageExtClass.method<&MessageExt::parseTopicFilterType>("parseTopicFilterType", { Php::ByVal("filterType", Php::Type::Numeric), });

    messageExtClass.method<&MessageExt::getStoreHostString>("getStoreHostString");

    messageExtClass.method<&MessageExt::getQueueId>("getQueue");
    messageExtClass.method<&MessageExt::setQueueId>("setQueueId", { Php::ByVal("queueId", Php::Type::Numeric), });

    messageExtClass.method<&MessageExt::getBornTimestamp>("getBornTimestamp");
    messageExtClass.method<&MessageExt::setBornTimestamp>("setBornTimestamp", { Php::ByVal("bornTimestamp", Php::Type::Numeric), });


    messageExtClass.method<&MessageExt::getMsgId>("getMsgId");
    messageExtClass.method<&MessageExt::setMsgId>("setMsgId", { Php::ByVal("msgId", Php::Type::String), });

    messageExtClass.method<&MessageExt::getOffsetMsgId>("getOffsetMsgId");
    messageExtClass.method<&MessageExt::setOffsetMsgId>("setOffsetMsgId", { Php::ByVal("offsetMsgId", Php::Type::Numeric), });

    messageExtClass.method<&MessageExt::getBodyCRC>("getBodyCRC");
    messageExtClass.method<&MessageExt::setBodyCRC>("setBodyCRC", { Php::ByVal("bodyCRC", Php::Type::Numeric), });

    messageExtClass.method<&MessageExt::getQueueOffset>("getQueueOffset");
    messageExtClass.method<&MessageExt::setQueueOffset>("setQueueOffset", { Php::ByVal("queueOffset", Php::Type::Numeric), });

    messageExtClass.method<&MessageExt::getCommitLogOffset>("getCommitLogOffset");
    messageExtClass.method<&MessageExt::setCommitLogOffset>("setCommitLogOffset", { Php::ByVal("physicOffset", Php::Type::Numeric), });

    messageExtClass.method<&MessageExt::getStoreSize>("getStoreSize");
    messageExtClass.method<&MessageExt::setStoreSize>("setStoreSize", { Php::ByVal("storeSize", Php::Type::Numeric), });

    messageExtClass.method<&MessageExt::getReconsumeTimes>("getReconsumeTimes");

    messageExtClass.method<&MessageExt::getPreparedTransactionOffset>("getPreparedTransactionOffset");
    messageExtClass.method<&MessageExt::setPreparedTransactionOffset>("setPreparedTransactionOffset", { Php::ByVal("preparedTransactionOffset", Php::Type::Numeric), });

    messageExtClass.method<&MessageExt::toString>("toString");
    messageExtClass.method<&MessageExt::getMessage>("getMessage");
    rocketMQNamespace.add(messageExtClass);
}
